### PR TITLE
Slashification of fun commands

### PR DESCRIPTION
**Pull request by sbrugel**

### CODE REVIEW

**Comment by joshualyon7** [comment]
Hey homie, PR this into slashCommands instead

**Comment by sbrugel** [comment]
Fixed :)

File: src/commands/admin/role.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+		const option: MessageSelectOptionData[] = [{
+			label: role.name,
+			value: role.id
+		}];
+
+		const menu = row.components[0] as MessageSelectMenu;
+		if (interaction.options.getString('mode') === 'Add') {
+			if (menu) {
+				for (let i = 0; i < menu.options.length; i++) {
```
Comment (joshualyon7) [inline src/commands/admin/role.ts:110]: A more javascript-y way to do this would be 
```
menu.options.forEach(option => {
    ...
});
```

File: src/commands/admin/role.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+		} else {
+			if (!menu) {
+				const responseEmbed = new MessageEmbed()
+					.setColor('#ff0000')
+					.setTitle('Argument error')
+					.setDescription(`This message does not have a role dropdown menu. You will need to create one using the /role command with 'Add' mode instead of 'Remove' mode.`);
+				return interaction.reply({ embeds: [responseEmbed], ephemeral: true });
+			}
+
+			for (let i = 0; i < menu.options.length; i++) {
```
Comment (joshualyon7) [inline src/commands/admin/role.ts:149]: same deal here

File: src/commands/fun/8ball.ts [hunk 3]
```diff
@@ -26,18 +26,32 @@ const MAGIC8BALL_RESPONSES = [
...
+		const response = question.length !== 0 && question[question.length - 1].endsWith('?')
+			?	MAGIC8BALL_RESPONSES[Math.floor(Math.random() * MAGIC8BALL_RESPONSES.length)]
+			:	'The 8-ball only responds to questions smh';
+		const responseEmbed = new MessageEmbed()
+			.setColor('#000000')
+			.setTitle('The magic 8-ball says...')
+			.setDescription(response)
+			.setFooter(`${interaction.user.username} asked: ${question}`);
+		return interaction.reply({ embeds: [responseEmbed], files: [{ attachment: `${__dirname}../../../../../assets/images/8-ball.png`, name: '8-ball.png' }] });
+		// if a user doesn't ask a question, should 'The 8-ball only responds to questions smh' be ephemeral?
```
Comment (joshualyon7) [inline src/commands/fun/8ball.ts:54]: Nah, it's definitely a feature of the command to publicly say that. @renross @bladetyr can comment if they want

File: src/commands/fun/f.ts [hunk 4]
```diff
@@ -1,16 +1,29 @@
...
 
 export default class extends Command {
 
-	description = 'Press F to pay respects';
-	aliases = ['respect'];
+	description = 'Press F to pay respects.';
+	options: ApplicationCommandOptionData[] = [
+		{
+			name: 'target',
+			description: '(Optional) The user to pay respects to',
```
Comment (joshualyon7) [inline src/commands/fun/f.ts:10]: I don't think this needs an `Optional` here, discord lets people know its optional

File: src/commands/fun/poll.ts [hunk 5]
```diff
@@ -1,111 +1,197 @@
...
+		},
+		{
+			name: 'question',
+			description: `What would you like to ask?`,
+			type: 'STRING',
+			required: true
+		},
+		{
+			name: 'choices',
+			description: `A poll can have 2-10 choices. Separate choices with '|' (no spaces/quotes).`,
```
Comment (joshualyon7) [inline src/commands/fun/poll.ts:28]: I personally think we should keep list arguments to comma-separated. @renross @bladetyr can weigh in though

File: src/commands/fun/poll.ts [hunk 6]
```diff
@@ -1,111 +1,197 @@
...
+				name: arg,
+				value: arg
+			}))
+		}
+	]
 	runInDM = false;
 
+	run(_msg: Message): Promise<void> { return; }
+
+	resetArray(array: number[], len: number): number[] {
```
Comment (joshualyon7) [inline src/commands/fun/poll.ts:47]: `Array.fill();` ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill)) accomplishes this in one line

File: src/commands/fun/poll.ts [hunk 7]
```diff
@@ -1,111 +1,197 @@
...
+			return interaction.reply({ embeds: [errorEmbed], ephemeral: true });
+		}
+
+		const mdTimestamp = `<t:${Math.floor(Date.now() / 1000) + (timespan / 1000)}:R>`;
 
 		let choiceText = '';
-		choices.forEach((choice, option) => {
-			choiceText += `${emotes[option]} ${choice}\n`;
-		});
+		for (let j = 0; j < choices.length; j++) {
```
Comment (joshualyon7) [inline src/commands/fun/poll.ts:97]: `choices.forEach(choice => { ... });`

File: src/commands/fun/poll.ts [hunk 8]
```diff
@@ -1,111 +1,197 @@
...
+		let pollEmbed = new MessageEmbed()
+			.setTitle(question)
+			.setDescription(`This poll was created by ${interaction.user.username} and ends **${mdTimestamp}**`)
+			.addField('Choices', choiceText)
+			.setFooter(pollFooter)
 			.setColor('RANDOM');
 
+		const choiceBtns = []; // first 5 choices
+		const choiceBtns2 = []; // next 5
+		for (let i = 0; i < choices.length; i++) {
```
Comment (joshualyon7) [inline src/commands/fun/poll.ts:114]: `choices.forEach(choice => { ... });`

**Review by joshualyon7 — CHANGES_REQUESTED** [review]

File: src/commands/fun/8ball.ts (hunk 3 above)
Comment (sbrugel) [reply src/commands/fun/8ball.ts:54]: Got it, was just curious.

File: src/commands/fun/f.ts (hunk 4 above)
Comment (sbrugel) [reply src/commands/fun/f.ts:10]: I simple put that there in case the "+1 optional" on Discord wasn't clear enough (since it is small print), but shall remove prior to the merge.

File: src/commands/fun/poll.ts (hunk 5 above)
Comment (sbrugel) [reply src/commands/fun/poll.ts:28]: This would mean if poll options use commas to, e.g. separate items, they would all be split between choices. What if someone wanted to make a poll where the choices were, say, "Eggs, ketchup, and cheese" and "Eggs, sausage, and bacon"? That would result in 6 choices not 2

File: src/commands/fun/8ball.ts (hunk 3 above)
Comment (BenSegal855) [reply src/commands/fun/8ball.ts:54]: It 100% was intended to be public.

File: src/commands/fun/poll.ts (hunk 5 above)
Comment (BenSegal855) [reply src/commands/fun/poll.ts:28]: Why not have four required args (question, time, option 1, option 2) and then a bunch of optional args for the rest of the options?

Comment (sbrugel) [reply src/commands/fun/poll.ts:28]: I was originally going to do something like that, but the issue that came to mind was if someone (somehow) skipped an optional choice while filling out the rest of their choices. Unless you were thinking of the code parsing each optional choice individually.

File: src/commands/admin/issue.ts [hunk 9]
```diff
@@ -8,6 +8,7 @@ import { Message } from 'discord.js';
 export default class extends Command {
 
 	usage = '<title>|[options]';
+	description = `Creates an issue in ${BOT.NAME}' repository.`;
```
[the remark this answers is not in the document]
Comment (sbrugel) [reply src/commands/admin/issue.ts]: Not too sure why this is here since I didn't edit this command?

File: src/commands/fun/poll.ts (hunk 5 above)
Comment (BenSegal855) [reply src/commands/fun/poll.ts:28]: You could pretty easily filter out options that are null and just show the ones that were given.

Comment (bladetyr) [reply src/commands/fun/poll.ts:28]: I like the options idea simply because that's more user friendly (discord tells the user that there are options that can be inputted) rather than someone doing "/poll red|green|blue," but not using options is faster than hitting tab to input poll options. I don't really like the comma split on the off chance that people put 'and' at the end (ex: 'red, green, and blue')

TLDR: I think the poll command is fine as-is, but making them options would be more user-friendly. 'red|green|blue' is faster to input though, so it's a matter of efficiency vs end user ease

Comment (sbrugel) [reply src/commands/fun/poll.ts:28]: That's what I was thinking, plus the instructions are clear enough, I'd like to think (saying to separate with |).

**Review by bladetyr — APPROVED** [review]
Check my comments under Josh's + Ben's conversation. I think either path you choose for the poll command will work perfectly fine, though anyone else's opinion is welcome :]

File: src/commands/admin/role.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+		if (!targetMessage) {
+			const responseEmbed = new MessageEmbed()
+				.setColor('#ff0000')
+				.setTitle('Argument error')
+				.setDescription(`Unknown message, make sure your channel and message ID are correct.`);
+			return interaction.reply({ embeds: [responseEmbed], ephemeral: true });
+		}
+
+		console.log(targetMessage.author.id);
+		console.log(BOT.CLIENT_ID);
```
Comment (BenSegal855) [inline src/commands/admin/role.ts:88]: Did you forget to take out some testing code?

File: src/commands/admin/role.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+	run(_msg: Message): Promise<void> { return; }
+
+	async tempRun(interaction: CommandInteraction): Promise<void> {
+		const channel = interaction.options.getChannel('channel') as TextChannel;
+		if (!channel || channel.type !== 'GUILD_TEXT') {
+			const responseEmbed = new MessageEmbed()
+				.setColor('#ff0000')
+				.setTitle('Argument error')
+				.setDescription(`You must tag a text channel.`);
+			return interaction.reply({ embeds: [responseEmbed], ephemeral: true });
```
Comment (BenSegal855) [inline src/commands/admin/role.ts:53]: You have this duplicated many times. Consider making it a function to condense the code a bit.

File: src/commands/fun/diceroll.ts [hunk 12]
```diff
@@ -28,29 +33,58 @@ export default class extends Command {
...
-			return DEFAULT_RANGE;
+		let responseEmbed: MessageEmbed;
+		if (!min) {
+			[min, max] = [DEFAULT_RANGE[0], DEFAULT_RANGE[1]];
+		} else if (!max && min) {
+			responseEmbed = new MessageEmbed()
+				.setColor('#ff0000')
+				.setTitle('Argument error')
+				.setDescription('If you provide a minimum, you must also provide a maximum.');
+			return interaction.reply({ embeds: [responseEmbed], ephemeral: true });
```
Comment (BenSegal855) [inline src/commands/fun/diceroll.ts:48]: Same deal as with the argument errors in the role command. You might want to make that function a util.

File: src/commands/partial visibility question/anonymous.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+		const courses: Array<Course> = await interaction.client.mongo.collection(DB.COURSES).find().toArray();
+
+		if (user.courses.length === 1) {
+			course = courses.find(c => c.name === user.courses[0]);
+		} else {
+			const inputtedCourse = courses.find(c => c.name === interaction.options.getString('course'));
+			if (!inputtedCourse) {
+				const responseEmbed = new MessageEmbed()
+					.setTitle(`Argument error`)
+					.setDescription('I wasn\'t able to determine your course based off of your enrollment or your input. Please specify the course at the beginning of your question.' +
```
Comment (BenSegal855) [inline src/commands/partial visibility question/anonymous.ts:53]: Shouldn't it be specified using the argument?

File: src/commands/submit.ts [hunk 14]
```diff
@@ -1,34 +1,87 @@
...
+		});
 	}
 
-	argParser(msg: Message, input: string): Array<string> {
-		if (msg.attachments.size !== 1) throw this.extendedHelp;
-		return [input];
+	countdown(interaction: CommandInteraction, timeout: number, embed: MessageEmbed): void {
+		const footerText = timeout > 1
+			? `You have ${timeout} seconds to submit.`
+			: `You have ${timeout} second to submit.`;
```
Comment (BenSegal855) [inline src/commands/submit.ts:82]:
```suggestion
		const footerText = `You have ${timeout} second${timeout !== 1 ? 's' : ''} to submit.`;
```

**Review by renross — APPROVED** [review]

File: src/commands/partial visibility question/anonymous.ts (hunk 13 above)
Comment (sbrugel) [reply src/commands/partial visibility question/anonymous.ts:53]: Unless I don't quite get what you're saying, this error message comes up if the user is enrolled in less/more than 1 class and does not specify the class they'd like to ask their question in (since the parameter is optional since Sage auto-detects your class if you're enrolled in just one).

If I've misinterpreted this comment, please correct me

File: src/commands/fun/diceroll.ts (hunk 12 above)
Comment (sbrugel) [reply src/commands/fun/diceroll.ts]: Done, and also integrated into the other commands that use this :)

File: src/pieces/commandManager.ts [hunk 15]
```diff
@@ -11,8 +13,51 @@ async function register(bot: Client): Promise<void> {
...
+						const course = courses.find(c => c.name === role.name.substring(5));
+						const user: SageUser = await interaction.client.mongo.collection(DB.USERS).findOne({ discordId: member.id });
+						user.courses = user.courses.filter(c => c !== course.name);
+						member.roles.remove(course.roles.student, `Unenrolled from ${course.name}.`);
+						member.roles.remove(id.value);
+						interaction.client.mongo.collection(DB.USERS).updateOne({ discordId: member.id }, { $set: { ...user } });
+						responseContent = `Your enrollments have been updated.`;
+					}
+				}
+				for (const id of values) {
```
[the remark this answers is not in the document]
Comment (sbrugel) [reply src/pieces/commandManager.ts]: Ren and I were messing around with converting this to a forEach loop the other day to no avail, mainly because the `continue` statement doesn't work in such a loop. We've tried different things, including replacing it with a `return` but problems kept popping up, e.g. if a user selected two roles, only one would be assigned to them

**Review by joshualyon7 — APPROVED** [review]
